cosigner/authcosigner.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 OpenPubkey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package cosigner
+
+import (
+	"crypto"
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type AuthCosigner struct {
+	Cosigner
+	Issuer         string
+	KeyID          string
+	AuthStateStore AuthStateStore
+}
+
+func New(signer crypto.Signer, alg jwa.SignatureAlgorithm, issuer, keyID string) (*AuthCosigner, error) {
+	hmacKey := make([]byte, 64)
+	if _, err := rand.Read(hmacKey); err != nil {
+		return nil, err
+	}
+
+	return &AuthCosigner{
+		Cosigner: Cosigner{
+			Alg:    alg,
+			Signer: signer},
+		Issuer:         issuer,
+		KeyID:          keyID,
+		AuthStateStore: NewAuthStateInMemoryStore(hmacKey),
+	}, nil
+}
+
+func (c *AuthCosigner) InitAuth(pkt *pktoken.PKToken, sig []byte) (string, error) {
```

Comment:
It looks like a lot of complex code is being added to support this version of PoP Auth. I strongly believe that this is something we should add after we have fully ratified a design. Especially since it's not a requirement of V1, I don't see a reason to rush it.

Reply:
This is just to authenticate the user to the cosigner. It is not the full POP Auth protocol from the paper as that requires Refresh and is general purpose.